### PR TITLE
added the docs directory with design report for kpm sparse checkout

### PR DESCRIPTION
**Pull request by d4v1d03**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

design doc for the issue #304

### CODE REVIEW

**Comment by d4v1d03** [comment]
I'll update it accordingly

**Comment by officialasishkumar** [comment]
Hey @Peefy 

>  other tools such as Cargo

Can you please elaborate on the tool in cargo which has similar functionality?

**Comment by d4v1d03** [comment]
yeah @Peefy , am really sorry, i'll take care for that in my next PR

File: docs/kpm_sparse_checkout.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+## KPM Sparse-Checkout feature
+
+## 1. Introduction
+kpm which is the package management tool for KCL, does not support sparse-checkout, which means an issue arises when dealing with monorepos, which may contain many KCL packages. We need to get a solution to download specific packages instead of all in a monorepo.
+
+## 2. Design research 
+A solution to this problem lies onto using GithubAPIs for kpm. The API can be used to fetch repository contents and then processing the response data to list the sub-directories of a specific monorepo. 
+
+## 3. User Interface
```
[the remark this answers is not in the document]
Comment (zong-zhe) [reply docs/kpm_sparse_checkout.md]: Sorry for the late review 😳, which is my mistake. I finished the review yesterday, but I forgot to click submit review. This review has been outdated.

[the remark this answers is not in the document]
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md]: @zong-zhe , please have a look

File: docs/kpm_sparse_checkout.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+## KPM Sparse-Checkout feature
```
[the remark this answers is not in the document]
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md:1]: done @Peefy

File: docs/kpm_sparse_checkout.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+This command will add the specified subdirectory as a dependency in the current KCL project.
+
+- Non-interactive Method :
+In addition to the command line interface, users can specify subdirectory dependencies directly in the `kcl.mod` file, which can be particularly useful for automated environments.
+
+## 4. Example Use-case
+Considering the nginx-ingres module, on typing the command 
+
+```
+kcl mod add https://github.com/kcl-lang/modules/tree/main/nginx-ingress /restrict-ingress-anotations
```
Comment (zong-zhe) [inline docs/kpm_sparse_checkout.md:27]: I just have a question, why don't you just merge these two parameters, something like this? 
```
kcl mod add https://github.com/kcl-lang/modules/tree/main/nginx-ingress/restrict-ingress-anotations
```
I'm not sure what the problems will be behind such a merging, so you need to consider it or have a research.

File: docs/kpm_sparse_checkout.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+Alternatively, for non-interactive environments, users can directly edit the `kcl.mod` file:
+```
+[package]
+name = "test"
+edition = "v0.9.0"
+version = "0.0.1"
+
+[dependencies]
+my_dependency = { git = "https://github.com/example/repo", subdirectory = "path/to/package" }
```
Comment (zong-zhe) [inline docs/kpm_sparse_checkout.md:40]: The problem here is the same as above.

```
[dependencies]
my_dependency = { git = "https://github.com/example/repo/path/to/package" }
```

Any problem like this ?

File: docs/kpm_sparse_checkout.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+## 6. Integration and the use of go-getter to download the specific subdirectories
```
Comment (zong-zhe) [inline docs/kpm_sparse_checkout.md:59]: This chapter I think could need more details.

File: docs/kpm_sparse_checkout.md (hunk 3 above)
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md:27]: Done the changes, i did some research and gave thought to it but i don't think there won't be much of any problem. If any problems arise during implementation, i'll adjust accordingly.

File: docs/kpm_sparse_checkout.md (hunk 4 above)
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md:40]: made the changes

File: docs/kpm_sparse_checkout.md (hunk 5 above)
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md:59]: Ok, I'll add some on my own, though can you still point out some specific details (if any) you're looking for.

**Comment by d4v1d03** [comment]
Sorry for the mess 😓, i just fixed my DCO

File: docs/kpm_sparse_checkout.md (hunk 5 above)
Comment (d4v1d03) [reply docs/kpm_sparse_checkout.md:59]: @zong-zhe

**Comment by d4v1d03** [comment]
@Peefy , @zong-zhe please ignore the mess, just fixed my DCO, done the doc changes as per zong-zhe suggested.

File: .github/dependabot.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# To get started with Dependabot version updates, you'll need to specify which
```
[the remark this answers is not in the document]
Comment (d4v1d03) [reply .github/dependabot.yaml]: Done @Peefy

**Comment by zong-zhe** [comment]
Hi @d4v1d03 😄

I've been looking at the cargo's design for this feature. I have some new findings to share with you. Hope this helpful.

I find this design comfortable to use. ❤️
It is here: https://doc.rust-lang.org/stable/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml

The `cargo.toml` looks like this:
```
[package]
name = "mypackage"
version = "0.0.1"

[dependencies]
bar = { git = "https://github.com/example/project", package = "foo" }
```

It will clone the git repo -> recursively look for all `cargo.toml` in the subdirectory -> find the package named `foo` -> load it.

I tried to use this feature on a project and found several advantages of this design.

- Compared to adding a subdirectory to the URL, this way the user does not need to bother typing the URL. The URL of the subdirectory in the browser is not directly usable.  😅

Screenshot 2024-06-06 at 11 13 02

- When a user decides to use a dependency, he should probably know the package name of the dependency. 😆

**Comment by d4v1d03** [comment]
Hey @zong-zhe , yes my research over the design i proposed was inspired from cargo's package managemnet. The same reason i was clarifying regarding the recursive check on all kcl.mod files. Thanks a lot for clarification over this.😊

**Pull request closed without merging**